src/main/java/com/iota/iri/hash/PearlDiver.java

```diff
@@ -26,15 +26,52 @@
     private volatile State state;
     private final Object syncObj = new Object();
 
+    public class Pow implements Runnable {
+        private final long[] midCurlStateLow;
+        private final long[] midCurlStateHigh;
+        private final int[] transactionTrits;
+        private final int mwm;
+
+        public Pow(final long[] midCurlStateLow, final long[] midCurlStateHigh, final int[] transactionTrits, final int mwm) {
+            this.midCurlStateLow = midCurlStateLow;
+            this.midCurlStateHigh = midCurlStateHigh;
+            this.transactionTrits = transactionTrits;
+            this.mwm = mwm;
+        }
+
+        @Override
+        public void run() {
+            long mask;
+            long outMask = 1;
+
+            if ((mask = PearlDiver.this.loop(this.midCurlStateLow, this.midCurlStateHigh, this.mwm)) != 0 &&
+                state == RUNNING) {
+                synchronized (syncObj) {
+                    state = COMPLETED;
+
+                    while ((outMask & mask) == 0) {
+                        outMask <<= 1;
+                    }
+
+                    for (int i = 0; i < CURL_HASH_LENGTH; i++) {
```

Comment:
This will also mutate the attachmentTimestamps? Please fix. You should only mutate the last 81 trits.